src/random.rs

```diff
@@ -0,0 +1,293 @@
+//!     let mut bitgen = default_bit_gen(py)?.lock()?;
+//!     // use bitgen without holding the GIL
+//!     Ok(py.allow_threads(|| bitgen.next_uint64()))
+//! })?;
+//! # Ok::<(), PyErr>(())
+//! ```
+//!
+//! With the [`rand`] crate installed, you can also use the [`rand::Rng`] APIs from the [`PyBitGeneratorGuard`]:
+//!
+//! ```
+//! # use pyo3::prelude::*;
+//! use rand::Rng as _;
+//! # use numpy::random::{PyBitGenerator, PyBitGeneratorMethods as _};
+//! # // TODO: reuse function definition from above?
+//! # fn default_bit_gen<'py>(py: Python<'py>) -> PyResult<Bound<'py, PyBitGenerator>> {
+//! #     let default_rng = py.import("numpy.random")?.call_method0("default_rng")?;
+//! #     let bit_generator = default_rng.getattr("bit_generator")?.downcast_into()?;
+//! #     Ok(bit_generator)
+//! # }
+//!
+//! Python::with_gil(|py| -> PyResult<_> {
+//!     let mut bitgen = default_bit_gen(py)?.lock()?;
+//!     if bitgen.random_ratio(1, 1_000_000) {
+//!         println!("a sure thing");
+//!     }
+//!     Ok(())
+//! })?;
+//! # Ok::<(), PyErr>(())
+//! ```
+//!
+//! [bg]: https://numpy.org/doc/stable//reference/random/bit_generators/generated/numpy.random.BitGenerator.html
+//! [ext]: https://numpy.org/doc/stable/reference/random/extending.html
+
+use std::ptr::NonNull;
+
+use pyo3::{
+    exceptions::PyRuntimeError,
+    ffi,
+    prelude::*,
+    sync::GILOnceCell,
+    types::{DerefToPyAny, PyCapsule, PyType},
+    PyTypeInfo,
+};
+
+use crate::npyffi::npy_bitgen;
+
+/// Wrapper for [`np.random.BitGenerator`][bg].
+///
+/// See also [`PyBitGeneratorMethods`].
+///
+/// [bg]: https://numpy.org/doc/stable//reference/random/bit_generators/generated/numpy.random.BitGenerator.html
+#[repr(transparent)]
+pub struct PyBitGenerator(PyAny);
+
+impl DerefToPyAny for PyBitGenerator {}
+
+unsafe impl PyTypeInfo for PyBitGenerator {
+    const NAME: &'static str = "PyBitGenerator";
+    const MODULE: Option<&'static str> = Some("numpy.random");
+
+    fn type_object_raw<'py>(py: Python<'py>) -> *mut ffi::PyTypeObject {
+        static CLS: GILOnceCell<Py<PyType>> = GILOnceCell::new();
+        let cls = CLS
+            .get_or_try_init::<_, PyErr>(py, || {
+                Ok(py
+                    .import("numpy.random")?
+                    .getattr("BitGenerator")?
+                    .downcast_into::<PyType>()?
+                    .unbind())
+            })
+            .expect("Failed to get BitGenerator type object")
+            .clone_ref(py)
+            .into_bound(py);
+        cls.as_type_ptr()
+    }
+}
+
+/// Methods for [`PyBitGenerator`].
+pub trait PyBitGeneratorMethods {
+    /// Acquire a lock on the BitGenerator to allow calling its methods in.
+    fn lock(&self) -> PyResult<PyBitGeneratorGuard>;
+}
+
+impl<'py> PyBitGeneratorMethods for Bound<'py, PyBitGenerator> {
+    fn lock(&self) -> PyResult<PyBitGeneratorGuard> {
+        let capsule = self.getattr("capsule")?.downcast_into::<PyCapsule>()?;
```

Comment:
You can use https://docs.rs/pyo3/latest/pyo3/macro.intern.html for the string literals here and elsewhere